src/net/metaplaylist/index.ts

```diff
@@ -0,0 +1,312 @@
+import { IKeySystem } from "../../parsers/manifest/types";
+import {
+    CustomSegmentLoader,
+    ILoaderObservable,
+    ILoaderResponse,
+    ImageParserObservable,
+    IManifestLoaderArguments,
+    IManifestParserArguments,
+    IManifestParserObservable,
+    IMetaTransportPipelines,
+    ISegmentLoaderArguments,
+    ISegmentParserArguments,
+    SegmentParserObservable,
+  } from "../types";
+
+import DASHTransport from "../dash";
+import SmoothTransport from "../smooth";
+
+interface IParserOptions {
+  segmentLoader? : CustomSegmentLoader;
+  suggestedPresentationDelay? : number;
+  referenceDateTime? : number;
+  minRepresentationBitrate? : number;
+  keySystems? : (hex? : Uint8Array) => IKeySystem[];
+}
+
+export interface IMetaManifestInfo {
+    manifests: Array<{
+      manifest: Document;
+      url: string;
+      startTime: number;
+      endTime: number;
+      transport: "dash"|"smooth";
+    }>;
+}
+
+type transportTypes = "dash"|"smooth";
+
+const transportTypes: transportTypes[] = ["dash", "smooth"];
+
+/**
+ * Parse playlist string to JSON.
+ * Returns an array of contents.
+ * @param {string} data
+ */
+function loadMetaPlaylistData(data: string): Array<{
+    url: string;
+    startTime: number;
+    endTime: number;
+    transport: "dash"|"smooth";
+  }> {
+  const parsedMetaPlaylist = JSON.parse(data);
+  const { contents } = parsedMetaPlaylist;
+  assert(contents);
+
+  return contents;
+}
+
+/**
+ * Get transport type from segment's privateInfos
+ * @param {Object} privateInfos
+ */
+function getTypeFromPrivateInfos(privateInfos: ISegmentPrivateInfos): transportTypes {
+  const transportType = privateInfos.manifestType;
+
+  if (!transportType) {
+    throw new Error("Undefined transport for content for metaplaylist.");
+  }
+
+  return transportType;
+}
+
+export default function(
+    options: IParserOptions = {}
+): IMetaTransportPipelines{
+
+    const transports = {
+      dash: DASHTransport(options),
+      smooth: SmoothTransport(options),
+    };
+
+    const manifestPipeline = {
+      loader(
+        { url } : IManifestLoaderArguments
+      ) : ILoaderObservable<IMetaManifestInfo> {
+        return request({
+          url,
+          responseType: "text",
+          ignoreProgressEvents: true,
+        }).mergeMap(({ value }) => {
+          const data = value.responseData;
+          const contents = loadMetaPlaylistData(data); // load data from meta document
+          const manifestsInfos$ = contents.map((content) => {
+            const transport = transports[content.transport];
+            const loader = transport.manifest.loader;
+
+            return loader({ url: content.url })
+              .filter((res) => res.type === "response")
+              .map((res) => {
+                return {
+                  // res has been filtered and is always a response event.
+                  // TS doesn't guess.
+                  manifest: (res as ILoaderResponse<Document>).value.responseData,
```

Comment:
I found a workaround for this, less dangerous than casting:
  - Define interfaces for every events emitted by the loader
  - in your filter, add a "is" type (ex: ``(res) : res is IMyEventA => res.type === "myEventA"``
  - in your map add the received type (ex: ``(res : IMyEventA) => {``

Should work and be future-proof.

Current solution would not work with HLS (as the loader do not return a document) and we wouldn't have any alert showing because of the casting here. Same case if we want a metametaplaylist